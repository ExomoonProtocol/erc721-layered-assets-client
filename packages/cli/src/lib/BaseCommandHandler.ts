/**
 * Base class for handling commands
 */
export class BaseCommandHandler {
  constructor() {}

  /**
   * Run the command
   */
  public async run() {
    throw new Error("Method not implemented.");
  }
}
