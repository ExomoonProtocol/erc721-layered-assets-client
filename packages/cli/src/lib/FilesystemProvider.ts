/* eslint-disable @typescript-eslint/no-explicit-any */

import { BaseFileProvider } from "@exomoon/erc721-layered-assets-client";
import path from "path";
import fs from "fs";

export class FilesystemProvider extends BaseFileProvider {
  constructor() {
    super();
  }

  public async fetchResource(
    _path: string,
    _options: Record<string, any> = {}
  ): Promise<any> {
    console.log(
      "[FilesystemProvider] Fetching resource",
      _path,
      typeof path,
      `<${path}>`
    );
    const absolutePath: fs.PathLike = path as any;

    console.log("Typeof path:", typeof absolutePath);

    if (!fs.existsSync(absolutePath)) {
      throw new Error(`File not found: ${absolutePath}`);
    }

    return fs.readFileSync(absolutePath, _options.encoding || "utf-8");
  }
}
