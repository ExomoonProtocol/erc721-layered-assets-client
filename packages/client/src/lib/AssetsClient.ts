/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  CollectionInfo,
  AssetsObject,
  Trait,
  ConditionalRenderingConfig,
} from "../models";
import axios from "axios";
import { ModelsUtils } from "../utils/ModelsUtils";

/**
 * Parameters for initializing a new AssetsClient instance.
 */
export interface AssetsClientInitParams {
  /**
   * Base URL to be used for fetching assets (N.B. the asets should be served as static files, and should be compliant with the Exomoon ERC721 Layered NFT standard)
   * It can be a link to a public S3 bucket or a web server you own.
   * @example "https://example.com"
   * @example "http://localhost:3000"
   **/
  baseUrl: string;

  /**
   * Specifies if client should use cache for fetching assets.
   * True by default.
   * @default true
   */
  useCache?: boolean;
}

/**
 * Base fetching parameters
 * (empty for now, will be used for future extensions)
 */
export interface BaseFetchingParams {}

/**
 * Assets client class. Represents a client for fetching assets.
 * It implements methods for fetching all the necessary assets about a Exomoon ERC721 Layered NFT collection: collection info, traits, variations, colors.
 * It also provides a caching mechanism for the fetched assets, in order to avoid fetching the same assets multiple times.
 *
 */
export class AssetsClient {
  private _baseUrl: string;

  private _useCache: boolean;

  private _cachedAssetsInfo: AssetsObject;

  constructor(initParams: AssetsClientInitParams) {
    this._baseUrl = initParams.baseUrl;
    this._useCache = initParams.useCache || false;

    this.initCache();
  }

  protected initCache() {
    this._cachedAssetsInfo = {};
  }

  protected async fetchResource(url: string): Promise<any> {
    console.log(`[AssetsClient] Fetching resource: ${url}`);
    const res = await axios.get(url);
    return res;
  }

  /**
   * Gets the base URL set during initialization
   * @returns Base URL string
   * @public
   */
  public get baseUrl(): string {
    return this._baseUrl;
  }

  protected validateParams(params: BaseFetchingParams): BaseFetchingParams {
    return {
      ...params,
    };
  }

  /**
   * Fetches collection info from remote server. If cache is enabled, it will return the cached collection info if it exists.
   * @param params Extra fetching parameters
   * @returns Collection info object, or null if not found
   * @public
   */
  public async fetchCollectionInfo(
    params?: BaseFetchingParams
  ): Promise<CollectionInfo | null> {
    const validatedParams = this.validateParams(params || {});

    let collectionInfoObj: CollectionInfo | null = null;

    if (this._useCache && this._cachedAssetsInfo?.collectionInfo) {
      collectionInfoObj = this._cachedAssetsInfo.collectionInfo;
    } else {
      const response = await this.fetchResource(
        `${this.baseUrl}/collection.json`
      );

      const rawObj =
        ModelsUtils.instance.serializer.deserializeObject<CollectionInfo>(
          response.data,
          CollectionInfo
        );

      console.log("[AssetsClient] Collection info raw:", rawObj);

      collectionInfoObj = rawObj || null;

      if (collectionInfoObj) {
        this._cachedAssetsInfo.collectionInfo = collectionInfoObj;
      }
    }

    return collectionInfoObj;
  }

  /**
   * Fetches traits info from remote server. If cache is enabled, it will return the cached traits if they exist.
   * @param params Extra fetching parameters
   * @returns Traits list array, or empty array if not found
   * @public
   */
  public async fetchTraits(params?: BaseFetchingParams): Promise<Array<Trait>> {
    const validatedParams = this.validateParams(params || {});
    const collectionInfo = await this.fetchCollectionInfo({});

    let traits: Array<Trait> = [];

    if (
      this._useCache &&
      this._cachedAssetsInfo?.traits &&
      this._cachedAssetsInfo.traits.length == collectionInfo?.traitsOrder.length
    ) {
      traits = this._cachedAssetsInfo.traits;
    } else {
      const traitNames = collectionInfo?.traitsOrder || [];
      traits = await Promise.all(
        traitNames.map(async (traitName) => {
          return await this.fetchTrait(traitName);
        })
      );

      if (traits.length > 0) {
        this._cachedAssetsInfo.traits = traits;
      }
    }

    return traits;
  }

  /**
   * Fetches a single trait from remote server. If cache is enabled, it will return the cached trait if it exists.
   * @param traitName Trait name
   * @param params Extra fetching parameters
   * @returns Trait object
   * @public
   */
  public async fetchTrait(
    traitName: string,
    params?: BaseFetchingParams
  ): Promise<Trait> {
    const validatedParams = this.validateParams(params || {});

    const response = await this.fetchResource(
      `${this.baseUrl}/traits/${traitName}/trait.json`
    );
    const rawObj = ModelsUtils.instance.serializer.deserializeObject<Trait>(
      response.data,
      Trait
    );

    if (rawObj) {
      // Cache trait
      if (this._cachedAssetsInfo.traits) {
        const traitIndex = this._cachedAssetsInfo.traits.findIndex(
          (t) => t.name === traitName
        );
        if (traitIndex !== -1) {
          this._cachedAssetsInfo.traits[traitIndex] = rawObj;
        } else {
          this._cachedAssetsInfo.traits.push(rawObj);
        }
      }

      return rawObj;
    } else {
      throw new Error("Trait not found");
    }
  }

  /**
   * Fetches all assets info from remote server. If cache is enabled, it will return the cached assets info if it exists.
   * The `AssetsObject` contains the collection info and the traits list, so it's a single object containing all the necessary assets info: after fetching this object, you can access all the other assets info through the cached object.
   * @param params Extra fetching parameters
   * @returns Assets object
   * @public
   */
  public async fetchAssetsObject(
    _params?: BaseFetchingParams
  ): Promise<AssetsObject> {
    const collectionInfo = await this.fetchCollectionInfo();
    const traits = await this.fetchTraits();

    this._cachedAssetsInfo.collectionInfo = collectionInfo!;
    this._cachedAssetsInfo.traits = traits;

    return this._cachedAssetsInfo;
  }

  /**
   * Get cached assets object. If no assets are cached, it will return null.
   * This method should be called after being sure that the assets have been fetched, otherwise it will return null.
   *
   * @returns Cached assets object
   * @public
   */
  public getAssetsObject(): AssetsObject | null {
    return this._cachedAssetsInfo || null;
  }

  /**
   * Get traits list. If no traits are cached, it will return an empty array.
   * This method should be called after being sure that the traits have been fetched, otherwise it will return an empty array.
   *
   * @returns Traits list array
   * @public
   */
  public getTrais(): Trait[] {
    return this._cachedAssetsInfo.traits || [];
  }

  /**
   * Get a single trait by name. If the trait is not cached, it will return null.
   * This method should be called after being sure that the trait has been fetched, otherwise it will return null.
   *
   * @param traitName Trait name
   * @returns Trait object, or null if not found
   * @public
   */
  public getTrait(traitName: string): Trait | null {
    if (this._cachedAssetsInfo.traits) {
      return (
        this._cachedAssetsInfo.traits.find((t) => t.name === traitName) || null
      );
    }

    return null;
  }

  /**
   * Get a single trait by index. If the trait is not cached, it will return null.
   * This method should be called after being sure that the trait has been fetched, otherwise it will return null.
   *
   * @param index Trait index
   * @returns Trait object, or null if not found
   * @public
   */
  public getTraitByIndex(index: number): Trait | null {
    if (this._cachedAssetsInfo.traits) {
      return this._cachedAssetsInfo.traits[index] || null;
    }

    return null;
  }

  /**
   * Get collection info. If no collection info is cached, it will return null.
   * This method should be called after being sure that the collection info has been fetched, otherwise it will return null.
   *
   * @returns Collection info object, or null if not found
   * @public
   */
  public getCollectionInfo(): CollectionInfo | null {
    return this._cachedAssetsInfo.collectionInfo || null;
  }

  /**
   * Clears the cache of fetched assets.
   * After calling this method, all the fetched assets will be removed from the cache, and the next time you call a method that fetches assets, it will fetch them again from the remote server.
   * @public
   */
  public clearCache(): void {
    this.initCache();
  }

  /**
   * Evals the image url for the provided trait, variation and color (optional).
   * If a conditional trait config is provided, it will override the default trait, variation and color, according to the conditional trait config itself.
   *
   * @param params Parameters for getting the image URL
   * @returns Image URL string
   * @public
   */
  public getTraitImageUrl(params: {
    traitName: string;
    variationName: string;
    colorName?: string;
    conditionalTraitConfig?: ConditionalRenderingConfig;
  }): string {
    const trait = this.getTrait(params.traitName);

    if (!trait) {
      throw new Error(`Trait ${params.traitName} not found`);
    }

    let traitUrlSection = trait.name;
    const variationUrlSection = params.variationName;
    const colorUrlSection = params.colorName
      ? params.colorName
      : params.variationName;

    if (params.conditionalTraitConfig) {
      const traits = this.getTrais();

      let configMatchedWithTrait = false;

      const config = params.conditionalTraitConfig;
      const trait = traits.find((t) => t.name === config.traitName);

      if (trait) {
        // If there is no variation name, use match only by trait name
        if (!config.variationName) {
          configMatchedWithTrait = true;
        }

        // If there is a variation name, match by trait name and variation name
        const variation = trait.variations.find(
          (v) => v.name === config.variationName
        );

        if (variation) {
          if (config.colorName) {
            // If there is a color name, match by trait name, variation name and color name
            const color = variation.colors.find((c) => c === config.colorName);

            if (color) {
              configMatchedWithTrait = true;
            }
          } else {
            // If there is no color name, match by trait name and variation name
            configMatchedWithTrait = true;
          }
        }

        if (configMatchedWithTrait) {
          traitUrlSection = `${params.traitName}/${config.folderName}`;
        }
      }
    }

    const imageUrl = `${this.baseUrl}/traits/${traitUrlSection}/${variationUrlSection}/${colorUrlSection}.png`;

    return imageUrl;
  }
}
