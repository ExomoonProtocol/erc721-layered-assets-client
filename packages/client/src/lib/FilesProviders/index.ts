export * from "./BaseFileProvider";
export * from "./HttpFileProvider";
