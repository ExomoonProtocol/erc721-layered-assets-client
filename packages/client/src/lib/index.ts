export * from "./AssetsClient";
export * from "./ItemConfiguration";
export * from "./TraitConfiguration";
