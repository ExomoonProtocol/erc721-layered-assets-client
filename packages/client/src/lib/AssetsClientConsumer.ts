import { AssetsClient } from "./AssetsClient";

/**
 * Base class for consumers of AssetsClient
 */
export class AssetsClientConsumer {
  private _assetsClient: AssetsClient;

  constructor(client: AssetsClient) {
    this._assetsClient = client;
  }

  /**
   * Get assets client instance used by the consumer
   */
  public get assetsClient(): AssetsClient {
    return this._assetsClient;
  }
}
