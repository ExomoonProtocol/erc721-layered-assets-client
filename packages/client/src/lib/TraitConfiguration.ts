import { AssetsClient } from "./AssetsClient";
import { AssetsClientConsumer } from "./AssetsClientConsumer";

export class TraitConfiguration extends AssetsClientConsumer {
  private _traitName: string;

  private _variationName: string;

  private _colorName?: string;

  constructor(_client: AssetsClient) {
    super(_client);
  }

  public get traitName(): string {
    return this._traitName;
  }

  public set traitName(value: string) {
    this._traitName = value;
  }

  public get variationName(): string {
    return this._variationName;
  }

  public set variationName(value: string) {
    this._variationName = value;
  }

  public get colorName(): string | undefined {
    return this._colorName;
  }

  public set colorName(value: string | undefined) {
    this._colorName = value;
  }

  public get metadataTraitName(): string {
    return this._traitName;
  }

  public get metadataTraitValue(): string {
    return `${this._variationName || ""}${
      this._colorName ? ` ${this._colorName}` : ""
    }`;
  }

  // public getImageUrl(): string {
  //   // return this.assetsClient.getTraitImageUrl(
  //   //   this._traitName,
  //   //   this._variationName,
  //   //   this._colorName
  //   // );
  // }
}
