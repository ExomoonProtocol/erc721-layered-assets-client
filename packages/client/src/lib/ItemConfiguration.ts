/* eslint-disable no-unused-vars */

import { AssetsClient } from "./AssetsClient";
import { AssetsClientConsumer } from "./AssetsClientConsumer";
import { TraitConfiguration } from "./TraitConfiguration";

export enum ItemConfigurationStatus {
  Unloaded = 0,
  Ready,
}

export class ItemConfiguration extends AssetsClientConsumer {
  private _traitConfigurations: Array<TraitConfiguration>;

  private _status: ItemConfigurationStatus;

  constructor(_client: AssetsClient) {
    super(_client);

    this._status = ItemConfigurationStatus.Unloaded;
    this._traitConfigurations = new Array<TraitConfiguration>();
  }

  /**
   * Prepare item configuration, and make sure all assets are loaded.
   */
  public async load(): Promise<void> {
    this._status = ItemConfigurationStatus.Ready;
    await this.assetsClient.fetchAssetsObject();
  }

  /**
   * Get trait configuration by trait name
   */
  public get traitConfigurations(): Array<TraitConfiguration> {
    return this._traitConfigurations;
  }

  /**
   * Get status of item configuration
   */
  public get status(): ItemConfigurationStatus {
    return this._status;
  }

  /**
   * Require item configuration to be ready
   */
  protected requireReady(): void {
    if (this._status !== ItemConfigurationStatus.Ready) {
      throw new Error("Item configuration not ready");
    }
  }

  /**
   * Set variation for a trait
   * @param trait Trait name
   * @param variation Variation name
   * @param color Color name
   */
  public setVariation(trait: string, variation: string, color?: string): void {
    this.requireReady();

    const traitConfiguration = this._traitConfigurations.find(
      (tc) => tc.traitName === trait
    );

    const traitObj = this.assetsClient.getTrait(trait);

    if (!traitObj) {
      throw new Error(`Trait ${trait} not found`);
    }

    const variationObj = traitObj.variations.find((v) => v.name === variation);

    if (!variationObj) {
      throw new Error(`Variation ${variation} not found`);
    }

    const variationHasColors =
      variationObj.colors && variationObj.colors.length > 0;

    let selectedColor: string | null = null;

    if (variationHasColors) {
      if (color) {
        selectedColor = variationObj.colors.find((c) => c === color) || null;

        if (!selectedColor) {
          throw new Error(`Color ${color} not found`);
        }
      } else {
        selectedColor = variationObj.colors[0];
      }
    } else if (color) {
      throw new Error("Variation doesn't support colors");
    }

    if (traitConfiguration) {
      traitConfiguration.variationName = variation;
      traitConfiguration.colorName = selectedColor || undefined;
    } else {
      const newTraitConfiguration = new TraitConfiguration(this.assetsClient);
      newTraitConfiguration.traitName = trait;
      newTraitConfiguration.variationName = variation;
      newTraitConfiguration.colorName = selectedColor || undefined;
      this._traitConfigurations.push(newTraitConfiguration);
      this.sortTraitConfigurations();
    }
  }

  /**
   * Sort trait configurations based on the order of traits in collection info
   */
  protected sortTraitConfigurations(): void {
    this._traitConfigurations = (
      this.assetsClient.getCollectionInfo()?.traitsOrder || []
    )
      .map((traitName) => {
        console.log(traitName);
        return this._traitConfigurations.find(
          (tc) => tc.traitName === traitName
        ) as TraitConfiguration;
      })
      .filter((tc) => tc);
  }

  /**
   * Get trait image urls
   * @returns Array of trait image urls
   */
  public getTraitsUrls(): Array<string> {
    this.requireReady();

    return this._traitConfigurations.map((tc) => {
      // Getting info about trait (asset info)
      const traitAsset = this.assetsClient.getTrait(tc.traitName);
      console.log("Asset: ", traitAsset);

      // Check if there is a conditional rendering config for this trait
      if (traitAsset?.conditonalRenderingConfig) {
        const conditionalRenderingConfig =
          traitAsset.conditonalRenderingConfig.find(
            (currentConditionalConfig) => {
              const matchedConfiguration = this._traitConfigurations.find(
                (tcToMatch) => {
                  let matched = true;

                  if (
                    currentConditionalConfig.traitName !== tcToMatch.traitName
                  ) {
                    matched = false;
                  } else if (
                    currentConditionalConfig.variationName &&
                    currentConditionalConfig.variationName !==
                      tcToMatch.variationName
                  ) {
                    matched = false;
                  } else if (
                    currentConditionalConfig.colorName &&
                    currentConditionalConfig.colorName !== tcToMatch.colorName
                  ) {
                    matched = false;
                  }

                  return matched;
                }
              );

              return matchedConfiguration;
            }
          );

        if (conditionalRenderingConfig) {
          console.log(conditionalRenderingConfig);
          return tc.getImageUrl(conditionalRenderingConfig);
        }
      }

      return tc.getImageUrl();
    });
  }
}
