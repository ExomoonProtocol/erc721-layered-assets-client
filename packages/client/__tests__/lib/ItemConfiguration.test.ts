import { AssetsClient } from "../../src/lib/AssetsClient";
import {
  ItemConfiguration,
  ItemConfigurationStatus,
} from "../../src/lib/ItemConfiguration";
import { HttpServerMock } from "../HttpServerMock";
import path from "path";

describe("ItemConfiguration class", () => {
  let client: AssetsClient;
  let client2: AssetsClient;
  let client3: AssetsClient;
  let client4: AssetsClient;

  beforeEach(() => {
    HttpServerMock.instance.clearTemporary();

    HttpServerMock.instance.addTemporaryResponseMappings([
      {
        url: "https://example.com",
        filePath: path.resolve(__dirname, "../assets/1/"),
      },
    ]);

    client = new AssetsClient({
      baseUrl: "https://example.com",
      useCache: true,
    });

    HttpServerMock.instance.addTemporaryResponseMappings([
      {
        url: "https://example2.com",
        filePath: path.resolve(__dirname, "../assets/2/"),
      },
    ]);

    client2 = new AssetsClient({
      baseUrl: "https://example2.com",
      useCache: true,
    });

    HttpServerMock.instance.addTemporaryResponseMappings([
      {
        url: "https://example3.com",
        filePath: path.resolve(__dirname, "../assets/3/"),
      },
    ]);

    client3 = new AssetsClient({
      baseUrl: "https://example3.com",
      useCache: true,
    });

    HttpServerMock.instance.addTemporaryResponseMappings([
      {
        url: "https://example4.com",
        filePath: path.resolve(__dirname, "../assets/4/"),
      },
    ]);

    client4 = new AssetsClient({
      baseUrl: "https://example4.com",
      useCache: true,
    });
  });

  describe("load method", () => {
    it("Should load item configuration", async () => {
      const itemConfiguration = new ItemConfiguration(client);
      await itemConfiguration.load();
      expect(itemConfiguration.status).toBe(ItemConfigurationStatus.Ready);
    });

    it("Should load initial item configuration", async () => {
      const itemConfiguration = new ItemConfiguration(client2);
      await itemConfiguration.load();
      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
      expect(itemConfiguration.status).toBe(ItemConfigurationStatus.Ready);
      expect(itemConfiguration.getTraitsUrls()).toMatchSnapshot();
    });

    it("Should get error if required trait is missing in initial configuration", async () => {
      const itemConfiguration = new ItemConfiguration(client4);

      await expect(itemConfiguration.load()).rejects.toThrow();
    });
  });

  describe("setVariation method", () => {
    it("Should set variation with a speciic color", async () => {
      const itemConfiguration = new ItemConfiguration(client);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Background", "Single Color", "Blue");
      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
    });

    it("Should set variation with color without specifying a color", async () => {
      const itemConfiguration = new ItemConfiguration(client);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Background", "Single Color");
      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
    });

    it("Should set variation with no color", async () => {
      const itemConfiguration = new ItemConfiguration(client);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Background", "Custom 1");
      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
    });

    it("Should throw an error if trying to set color for a variation that doesn't support it", async () => {
      const itemConfiguration = new ItemConfiguration(client);
      await itemConfiguration.load();
      expect(() =>
        itemConfiguration.setVariation("Background", "Custom 1", "Blue")
      ).toThrow();
    });
  });

  describe("removeVariation method", () => {
    it("Should remove variation", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Background", "Single Color", "Blue");
      itemConfiguration.setVariation("Cover", "Custom Cover 1");

      itemConfiguration.removeVariation("Cover");
      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
    });

    it("Should throw an error if trying to remove a required variation", async () => {
      const itemConfiguration = new ItemConfiguration(client2);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Background", "Single Color", "Blue");

      expect(() => itemConfiguration.removeVariation("Background")).toThrow();
    });

    it("Should throw an error if trying to remove a variation that doesn't exist", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      expect(() => itemConfiguration.removeVariation("Cover")).toThrow();
      expect(() =>
        itemConfiguration.removeVariation("Unknown Trait")
      ).toThrow();
    });
  });

  describe("getTraitsUrls method", () => {
    it("Should get trait urls - single trait", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Shape", "Circle");
      expect(itemConfiguration.getTraitsUrls()).toMatchSnapshot();
    });

    it("Should get trait urls - multiple traits", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Shape", "Circle");
      itemConfiguration.setVariation("Background", "Single Color", "Blue");
      expect(itemConfiguration.getTraitsUrls()).toMatchSnapshot();
    });

    it("Should get trait urls - multiple traits with conditional rendering", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Shape", "Circle");
      itemConfiguration.setVariation("Background", "Single Color", "Blue");
      itemConfiguration.setVariation("Cover", "Custom Cover 1");
      expect(itemConfiguration.getTraitsUrls()).toMatchSnapshot();
    });
  });

  describe("renderMetadata method", () => {
    it("Should render metadata", async () => {
      const itemConfiguration = new ItemConfiguration(client3);
      await itemConfiguration.load();
      itemConfiguration.setVariation("Shape", "Circle");
      itemConfiguration.setVariation("Background", "Single Color", "Blue");
      itemConfiguration.setVariation("Cover", "Custom Cover 1");
      expect(itemConfiguration.renderMetadata().serialize()).toMatchSnapshot();
    });
  });

  describe("buildFromLayersDataString method", () => {
    it("Should build item configuration from layers data string", async () => {
      const itemConfiguration =
        await ItemConfiguration.buildFromLayersDataString("0x00", client3);

      expect(itemConfiguration.traitConfigurations).toMatchSnapshot();
    });
  });
});
