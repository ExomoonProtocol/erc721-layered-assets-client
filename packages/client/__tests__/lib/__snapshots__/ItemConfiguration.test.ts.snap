// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemConfiguration class buildFromLayersDataString method Should build item configuration from layers data string 1`] = `
[
  TraitConfiguration {
    "_assetsClient": AssetsClient {
      "_baseUrl": "https://example2.com",
      "_cachedAssetsInfo": {
        "collectionInfo": CollectionInfo {
          "collectionName": "Collection 1",
          "traitsOrder": [
            "Background",
            "Shape",
            "Cover",
          ],
        },
        "traits": [
          Trait {
            "iconUrl": "https://example2.com/traits/Background/icon.svg",
            "name": "Background",
            "variations": [
              Variation {
                "colors": [
                  "Black",
                  "Blue",
                  "Brown",
                  "Green",
                ],
                "name": "Single Color",
                "previewImageUrl": "https://example2.com/traits/Background/thumbnail.png",
              },
              Variation {
                "name": "Custom 1",
                "previewImageUrl": "https://example2.com/traits/Background/thumbnail.png",
              },
            ],
          },
          Trait {
            "iconUrl": "https://example2.com/traits/Shape/icon.svg",
            "name": "Shape",
            "variations": [
              Variation {
                "name": "Rectangle",
                "previewImageUrl": "https://example2.com/traits/Shape/thumbnail.png",
              },
              Variation {
                "name": "Circle",
                "previewImageUrl": "https://example2.com/traits/Shape/thumbnail.png",
              },
              Variation {
                "name": "Triangle",
                "previewImageUrl": "https://example2.com/traits/Shape/thumbnail.png",
              },
            ],
          },
          Trait {
            "conditonalRenderingConfig": [
              ConditionalRenderingConfig {
                "colorName": undefined,
                "traitName": "Shape",
                "variationName": "Rectangle",
              },
              ConditionalRenderingConfig {
                "colorName": undefined,
                "traitName": "Shape",
                "variationName": "Circle",
              },
              ConditionalRenderingConfig {
                "colorName": undefined,
                "traitName": "Shape",
                "variationName": "Triangle",
              },
            ],
            "iconUrl": "https://example2.com/traits/Cover/icon.svg",
            "name": "Cover",
            "variations": [
              Variation {
                "colors": [
                  "Red",
                  "Blue",
                ],
                "name": "Custom Cover 1",
                "previewImageUrl": "https://example2.com/traits/Cover/thumbnail.png",
              },
            ],
          },
        ],
      },
      "_fileProvider": HttpFileProvider {},
      "_useCache": true,
    },
    "_colorName": "Black",
    "_traitName": "Background",
    "_variationName": "Single Color",
  },
]
`;

exports[`ItemConfiguration class getTraitsUrls method Should get trait urls - multiple traits 1`] = `
[
  "https://example2.com/traits/Background/Single Color/Blue.png",
  "https://example2.com/traits/Shape/Circle/Circle.png",
]
`;

exports[`ItemConfiguration class getTraitsUrls method Should get trait urls - multiple traits with conditional rendering 1`] = `
[
  "https://example2.com/traits/Background/Single Color/Blue.png",
  "https://example2.com/traits/Shape/Circle/Circle.png",
  "https://example2.com/traits/Cover/Shape_Circle/Custom Cover 1/Red.png",
]
`;

exports[`ItemConfiguration class getTraitsUrls method Should get trait urls - single trait 1`] = `
[
  "https://example2.com/traits/Shape/Circle/Circle.png",
]
`;

exports[`ItemConfiguration class renderMetadata method Should render metadata 1`] = `
{
  "attributes": [
    {
      "trait_type": "Background",
      "value": "Single Color Blue",
    },
    {
      "trait_type": "Shape",
      "value": "Circle",
    },
    {
      "trait_type": "Cover",
      "value": "Custom Cover 1 Red",
    },
  ],
  "description": "",
  "image": "{{imageUrl}}",
  "name": "Collection 1 #{{id}}",
}
`;

exports[`ItemConfiguration class setVariation method Should set variation with a speciic color 1`] = `
[
  TraitConfiguration {
    "_assetsClient": AssetsClient {
      "_baseUrl": "https://example.com",
      "_cachedAssetsInfo": {
        "collectionInfo": CollectionInfo {
          "collectionName": "Collection 1",
          "traitsOrder": [
            "Background",
          ],
        },
        "traits": [
          Trait {
            "description": "A simple background layer",
            "iconUrl": "https://example.com/traits/Background/icon.svg",
            "name": "Background",
            "variations": [
              Variation {
                "colors": [
                  "Black",
                  "Blue",
                  "Brown",
                  "Green",
                ],
                "description": "A single color background",
                "name": "Single Color",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "common",
              },
              Variation {
                "description": "A custom background",
                "name": "Custom 1",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "rare",
              },
            ],
          },
        ],
      },
      "_fileProvider": HttpFileProvider {},
      "_useCache": true,
    },
    "_colorName": "Blue",
    "_traitName": "Background",
    "_variationName": "Single Color",
  },
]
`;

exports[`ItemConfiguration class setVariation method Should set variation with color without specifying a color 1`] = `
[
  TraitConfiguration {
    "_assetsClient": AssetsClient {
      "_baseUrl": "https://example.com",
      "_cachedAssetsInfo": {
        "collectionInfo": CollectionInfo {
          "collectionName": "Collection 1",
          "traitsOrder": [
            "Background",
          ],
        },
        "traits": [
          Trait {
            "description": "A simple background layer",
            "iconUrl": "https://example.com/traits/Background/icon.svg",
            "name": "Background",
            "variations": [
              Variation {
                "colors": [
                  "Black",
                  "Blue",
                  "Brown",
                  "Green",
                ],
                "description": "A single color background",
                "name": "Single Color",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "common",
              },
              Variation {
                "description": "A custom background",
                "name": "Custom 1",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "rare",
              },
            ],
          },
        ],
      },
      "_fileProvider": HttpFileProvider {},
      "_useCache": true,
    },
    "_colorName": "Black",
    "_traitName": "Background",
    "_variationName": "Single Color",
  },
]
`;

exports[`ItemConfiguration class setVariation method Should set variation with no color 1`] = `
[
  TraitConfiguration {
    "_assetsClient": AssetsClient {
      "_baseUrl": "https://example.com",
      "_cachedAssetsInfo": {
        "collectionInfo": CollectionInfo {
          "collectionName": "Collection 1",
          "traitsOrder": [
            "Background",
          ],
        },
        "traits": [
          Trait {
            "description": "A simple background layer",
            "iconUrl": "https://example.com/traits/Background/icon.svg",
            "name": "Background",
            "variations": [
              Variation {
                "colors": [
                  "Black",
                  "Blue",
                  "Brown",
                  "Green",
                ],
                "description": "A single color background",
                "name": "Single Color",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "common",
              },
              Variation {
                "description": "A custom background",
                "name": "Custom 1",
                "previewImageUrl": "https://example.com/traits/Background/thumbnail.png",
                "rarity": "rare",
              },
            ],
          },
        ],
      },
      "_fileProvider": HttpFileProvider {},
      "_useCache": true,
    },
    "_colorName": undefined,
    "_traitName": "Background",
    "_variationName": "Custom 1",
  },
]
`;
